### PR TITLE
(maint) allow puppet server installs from version

### PR DESCRIPTION
**Pull request by er0ck**
Without this change the pre-suite installs puppet-server from
    nightlies/puppetserver-latest which can sometimes be an older version.
    this fix allows one to specify SERVER_VERSION environment variable to
    pull a specific build.
    This change also adds puppetserver-confdir to the aio options file.

### CODE REVIEW

File: acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb [hunk 1]
```diff
@@ -4,10 +4,10 @@
 
 step "(PUP-4004) Set permissions on puppetserver directories that currently live in the agent cache dir"
 %w[reports server_data yaml bucket].each do |dir|
-  on master, "install --directory /opt/puppetlabs/puppet/cache/#{dir}"
+  on master, "install --directory #{options[:HOSTS][:master][:puppetvardir]}/#{dir}"
```
Comment (joshcooper) [inline acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb]: Ideally, I hope we can do:

```
vardir = master.puppet('master')['vardir']
```
This will effectively run `puppet master --configprint vardir` on the master node, so that we don't need to rely on `vardir` being set in the host yaml files (or beaker). This assumes that the master is setting the correct values in the shared `puppet.conf` and not in some other config file that puppet (ruby) can't see.

File: acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb [hunk 2]
```diff
@@ -34,10 +34,10 @@
 # of the default master-var-dir in puppetserver, and the user
 # based codedir.
 step "(SERVER-347) Set required codedir setting on puppetserver"
-on master, puppet("config set codedir /etc/puppetlabs/code --section master")
+on master, puppet("config set codedir #{options[:HOSTS][:master][:puppetcodedir]} --section master")
```
Comment (joshcooper) [inline acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb]: same comment here about `codedir`

File: acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb [hunk 3]
```diff
@@ -34,10 +34,10 @@
...
 
 step "(SERVER-370) overwrite ruby-load-path"
-create_remote_file(master, '/etc/puppetserver/conf.d/os-settings.conf', <<-EOF)
+create_remote_file(master, "#{options[:puppetserver_confdir]}/os-settings.conf", <<-EOF)
```
Comment (joshcooper) [inline acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb]: :+1: Unfortunately I don't think this setting is accessible via (ruby) puppet

**Comment by er0ck** [comment]
thanks @joshcooper.  fixed-up

File: acceptance/config/aio/options.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   :type => 'aio',
   :is_puppetserver => true,
+  :puppetserver_confdir => '/etc/puppetlabs/puppetserver/conf.d',
```
Comment (joshcooper) [inline acceptance/config/aio/options.rb]: If we merge this now, then AIO acceptance will fail running against the *current* nightly puppet-server, until builds with https://tickets.puppetlabs.com/browse/SERVER-387 are promoted to nightly. Can we leave this as `/etc/puppetserver/conf.d` for now? Then when the new puppet-server is promoted, we'll encounter failures on our end telling us to change `puppetserver_confdir`?

Comment (joshcooper) [inline acceptance/config/aio/options.rb]: Also I just remembered that beaker uses `puppetserver_confdir` when bouncing the puppetserver: https://github.com/puppetlabs/beaker/blob/beaker2.5.0/lib/beaker/dsl/helpers.rb#L551. Will moving this to the options file instead of the host files cause beaker to fall back to its hardcoded (and incorrect) AIO defaults?

**Comment by joshcooper** [comment]
@er0ck can you rebase your PR on master? We recently removed the workaround that overrode `puppetserver_confdir` as the puppetserver packages are doing the right thing now, so you'll be able to remove the change to `options.rb` and various host config files.

File: acceptance/config/aio/options.rb (hunk 4 above)
Comment (er0ck) [inline acceptance/config/aio/options.rb]: tested this against various tests that use with_puppet_running_on, and they all work fine.  the options hash is the options hash if it comes from any of the 6! places you can set options

**Comment by er0ck** [comment]
@joshcooper rebased against master and tested against rhel7 and win2012

File: acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb [hunk 5]
```diff
@@ -1,9 +1,10 @@
 step "(PUP-4004) Set permissions on puppetserver directories that currently live in the agent cache dir"
+vardir = master.puppet('master')['vardir']
 %w[reports server_data yaml bucket].each do |dir|
-  on master, "install --directory /opt/puppetlabs/puppet/cache/#{dir}"
+  on master, "install --directory #{vardir}/#{dir}"
 end
-on master, "chown -R puppet:puppet /opt/puppetlabs/puppet/cache"
-on master, "chmod -R 750 /opt/puppetlabs/puppet/cache"
+on master, "chown -R puppet:puppet #{vardir}"
+on master, "chmod -R 750 #{vardir}"
```
Comment (joshcooper) [inline acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb]: @er0ck could you remove this part, this workaround was removed in  https://github.com/puppetlabs/puppet/commit/482250f2c3032d483565755566ad1e695521d4f1

File: acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb [hunk 6]
```diff
@@ -1,10 +1,3 @@
-step "(PUP-4004) Set permissions on puppetserver directories that currently live in the agent cache dir"
-%w[reports server_data yaml bucket].each do |dir|
-  on master, "install --directory /opt/puppetlabs/puppet/cache/#{dir}"
-end
-on master, "chown -R puppet:puppet /opt/puppetlabs/puppet/cache"
-on master, "chmod -R 750 /opt/puppetlabs/puppet/cache"
-
```
Comment (joshcooper) [inline acceptance/setup/aio/pre-suite/020_AIO_Workarounds.rb]: @er0ck can you undo this change (as it's already been done in https://github.com/puppetlabs/puppet/commit/0c138daad9ba0ed595956a2ba046d0cab08756ef), and update the commit message?

File: acceptance/config/aio/options.rb (hunk 4 above)
Comment (joshcooper) [inline acceptance/config/aio/options.rb]: ah, this actually need to be  `:'puppetserver-confdir' => '/etc/puppetlabs/puppetserver/conf.d',` note the dash, which has to be quoted since it's a symbol. I think it was passing for you because the version of beaker we are pinned to already set that as the AIO default.

Let's hold off on this until we get puppet testing against beaker#master

Comment (er0ck) [inline acceptance/config/aio/options.rb]: weird. i thought beaker converted the dashes in arguments and env vars to underscores for use in its options hash. but this is just an inconsistent convention...

**Comment by joshcooper** [comment]
@er0ck the options.rb and node definition files have been overtaken by events leading up to rc1. Could you update the commit to only contain the `010_Install.rb` change?

**Comment by er0ck** [comment]
@joshcooper do we really want to revert to getting puppetserver-confdir from beaker?  
i suppose that setting really needs to live with the puppet-server code and not with puppet, so i suppose i should remove it from options.rb for that reason alone.
is there a way to query puppet-server for its confdir after its installed?  not sure if that will work in all cases.

**Comment by joshcooper** [comment]
looks good, will merge after rc1 freeze
